pkg/ingester/ingester.go

```diff
@@ -1005,63 +1032,83 @@ func (i *Ingester) startPushRequest(ctx context.Context, reqSize int64) (_ conte
 			if (rejectEqualInflightBytes && inflightBytes >= il.MaxInflightPushRequestsBytes) || inflightBytes > il.MaxInflightPushRequestsBytes {
 				i.metrics.rejected.WithLabelValues(reasonIngesterMaxInflightPushRequestsBytes).Inc()
-				return nil, false, errMaxInflightRequestsBytesReached
+				return nil, false, acquiredCircuitBreakerPermit, errMaxInflightRequestsBytesReached
 			}
 		}
 
 		if il.MaxIngestionRate > 0 {
 			if rate := i.ingestionRate.Rate(); rate >= il.MaxIngestionRate {
 				i.metrics.rejected.WithLabelValues(reasonIngesterMaxIngestionRate).Inc()
-				return nil, false, errMaxIngestionRateReached
+				return nil, false, acquiredCircuitBreakerPermit, errMaxIngestionRateReached
 			}
 		}
 	}
 
 	finishRequestInDefer = false
-	return ctx, true, nil
+	return ctx, true, acquiredCircuitBreakerPermit, nil
 }
 
-func (i *Ingester) finishPushRequest(reqSize int64) {
+func (i *Ingester) finishPushRequest(ctx context.Context, reqSize int64, duration time.Duration, acquiredCircuitBreakerPermit bool, err error) {
 	i.inflightPushRequests.Dec()
 	if reqSize > 0 {
 		i.inflightPushRequestsBytes.Sub(reqSize)
 	}
+	if acquiredCircuitBreakerPermit {
+		_ = i.circuitBreaker.finishPushRequest(ctx, duration, err)
+	}
 }
 
 // PushWithCleanup is the Push() implementation for blocks storage and takes a WriteRequest and adds it to the TSDB head.
-func (i *Ingester) PushWithCleanup(ctx context.Context, req *mimirpb.WriteRequest, cleanUp func()) error {
+func (i *Ingester) PushWithCleanup(ctx context.Context, req *mimirpb.WriteRequest, cleanUp func()) (err error) {
 	// NOTE: because we use `unsafe` in deserialisation, we must not
 	// retain anything from `req` past the exit from this function.
 	defer cleanUp()
 
+	start := time.Now()
 	// Only start/finish request here when the request comes NOT from grpc handlers (i.e., from ingest.Store).
 	// NOTE: request coming from grpc handler may end up calling start multiple times during its lifetime (e.g., when migrating to ingest storage).
 	// startPushRequest handles this.
 	if i.cfg.IngestStorageConfig.Enabled || !i.cfg.LimitInflightRequestsUsingGrpcMethodLimiter {
 		reqSize := int64(req.Size())
-		_, shouldFinish, err := i.startPushRequest(ctx, reqSize)
-		if err != nil {
-			return middleware.DoNotLogError{Err: err}
+		_, shouldFinish, acquiredCircuitBreakerPermit, startPushErr := i.startPushRequest(ctx, reqSize)
+		if startPushErr != nil {
+			return middleware.DoNotLogError{Err: startPushErr}
 		}
 		if shouldFinish {
-			defer i.finishPushRequest(reqSize)
+			defer func() {
+				i.finishPushRequest(ctx, reqSize, time.Since(start), acquiredCircuitBreakerPermit, err)
+			}()
 		}
 	}
 
+	defer func() {
+		// We enrich the pushRequestState contained in the context with this PushWithCleanUp()
+		// call duration, and a possible error it returns. These data are needed during a
+		// successive call to FinishPushRequest().
+		if st, ok := ctx.Value(pushReqCtxKey).(*pushRequestState); ok {
```

Comment:
Note that for this to work, `ctx` must contain the `pushRequestState`. If we called `startPushRequest` in this method, then we must keep returned context and use that!